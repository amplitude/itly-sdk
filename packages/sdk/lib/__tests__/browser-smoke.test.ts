/**
 * Test for Browser version of SDK
 * @jest-environment jsdom
 */
/* eslint-disable import/no-unresolved, global-require, import/extensions, no-unused-vars */
import CustomPlugin from '../../../../__tests__/src/CustomPlugin';
import ErrorPlugin from '../../../../__tests__/src/ErrorPlugin';
import requireForTestEnv from '../../../../__tests__/util/requireForTestEnv';
import DummyPlugin from '../../../../__tests__/src/DummyPlugin';

beforeEach(() => {
  jest.resetModules();
});

test('should load and track events to a custom destination (no validation)', () => {
  const consoleSpy = jest.spyOn(console, 'log').mockImplementation();
  const userId = 'test-user-id';

  const itly = requireForTestEnv(__dirname);

  itly.load({
    environment: 'production',
    context: {
      requiredString: 'A required string',
      optionalEnum: 'Value 1',
    },
    plugins: [new CustomPlugin()],
  });

  itly.identify(undefined, {
    userProp: 1,
  });

  itly.alias(userId);

  itly.group('a-group-id', {
    groupProp: 'test value',
  });

  itly.page('page category', 'page name', {
    pageProp: 'a page property',
  });

  itly.track({
    name: 'Event No Properties',
  });

  itly.track({
    name: 'Event With All Properties',
    properties: {
      requiredString: 'A required string',
      requiredNumber: 2.0,
      requiredInteger: 42,
      requiredEnum: 'Enum1',
      requiredBoolean: false,
      requiredConst: 'some-const-value',
      requiredArray: ['required', 'array'],
      optionalString: 'I\'m optional!',
    },
  });

  try {
    itly.track({
      name: 'EventMaxIntForTest',
      properties: {
        intMax10: 20,
      },
    });
  } catch (e) {
    // do nothing
  }

  const customOnly = itly.getPlugin('custom');
  // eslint-disable-next-line no-console
  console.log('CustomPlugin.id()', customOnly!.id());

  expect(consoleSpy.mock.calls).toMatchSnapshot();

  consoleSpy.mockRestore();
});

test('other plugins should continue if another plugin throws errors in callback methods', () => {
  const consoleSpy = jest.spyOn(console, 'log').mockImplementation();
  const itly = requireForTestEnv(__dirname);

  const id = 'user-id';
  const dummyPlugin = new DummyPlugin();
  const methodsToSpyOn: string[] = [
    'load',
    'alias',
    'identify', 'group', 'page', 'track',
    'postIdentify', 'postGroup', 'postTrack',
    'reset',
  ];

  const pluginSpies: { [key: string]: jest.SpyInstance } = {};
  methodsToSpyOn.forEach((method) => {
    pluginSpies[method] = jest.spyOn(dummyPlugin, method as any);
  });

  itly.load({
    environment: 'production',
    plugins: [
      new ErrorPlugin(),
      dummyPlugin,
    ],
  });
  itly.identify('tmp-id');
  itly.alias(id, 'tmp-id');
  itly.group('group-id');
  itly.page('category', 'name');
  itly.track({
    name: 'my-event',
    properties: {
      prop: 'a value',
    },
  });
  itly.reset();

  methodsToSpyOn.forEach((method) => {
    expect(pluginSpies[method]).toHaveBeenCalledTimes(1);
  });

  expect(consoleSpy.mock.calls).toMatchSnapshot();
  consoleSpy.mockRestore();
});
