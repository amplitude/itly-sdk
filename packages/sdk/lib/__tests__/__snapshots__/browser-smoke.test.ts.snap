// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`other plugins should continue if another plugin throws errors in callback methods 1`] = `Array []`;

exports[`should load and track events to a custom destination (no validation) 1`] = `
Array [
  Array [
    "CustomPlugin: ",
    "load() environment='production'",
  ],
  Array [
    "CustomPlugin: ",
    "validate() event='identify' properties={\\"userProp\\":1}",
  ],
  Array [
    "CustomPlugin: ",
    "identify() userId='undefined' properties={\\"userProp\\":1}",
  ],
  Array [
    "CustomPlugin: ",
    "postIdentify() userId='undefined' properties={\\"userProp\\":1} validationResponses=[{\\"valid\\":true,\\"pluginId\\":\\"custom\\"}]",
  ],
  Array [
    "CustomPlugin: ",
    "alias() userId='test-user-id' previousId='undefined'",
  ],
  Array [
    "CustomPlugin: ",
    "validate() event='group' properties={\\"groupProp\\":\\"test value\\"}",
  ],
  Array [
    "CustomPlugin: ",
    "group() userId='undefined' groupId='a-group-id' properties={\\"groupProp\\":\\"test value\\"}",
  ],
  Array [
    "CustomPlugin: ",
    "postGroup() userId='undefined' groupId='a-group-id' properties={\\"groupProp\\":\\"test value\\"} validationResponses=[{\\"valid\\":true,\\"pluginId\\":\\"custom\\"}]",
  ],
  Array [
    "CustomPlugin: ",
    "validate() event='page' properties={\\"pageProp\\":\\"a page property\\"}",
  ],
  Array [
    "CustomPlugin: ",
    "page() userId='undefined' category='page category' name='page name' properties={\\"pageProp\\":\\"a page property\\"}",
  ],
  Array [
    "CustomPlugin: ",
    "postPage() userId='undefined' category='page category' name='page name' properties={\\"pageProp\\":\\"a page property\\"} validationResponses=[{\\"valid\\":true,\\"pluginId\\":\\"custom\\"}]",
  ],
  Array [
    "CustomPlugin: ",
    "validate() event='Event No Properties' properties=undefined",
  ],
  Array [
    "CustomPlugin: ",
    "validate() event='context' properties={\\"requiredString\\":\\"A required string\\",\\"optionalEnum\\":\\"Value 1\\"}",
  ],
  Array [
    "CustomPlugin: ",
    "track() userId='undefined' event='Event No Properties' properties={\\"requiredString\\":\\"A required string\\",\\"optionalEnum\\":\\"Value 1\\"}",
  ],
  Array [
    "CustomPlugin: ",
    "postTrack() userId='undefined' event='Event No Properties' properties={\\"requiredString\\":\\"A required string\\",\\"optionalEnum\\":\\"Value 1\\"} validationResponses=[{\\"valid\\":true,\\"pluginId\\":\\"custom\\"},{\\"valid\\":true,\\"pluginId\\":\\"custom\\"}]",
  ],
  Array [
    "CustomPlugin: ",
    "validate() event='Event With All Properties' properties={\\"requiredString\\":\\"A required string\\",\\"requiredNumber\\":2,\\"requiredInteger\\":42,\\"requiredEnum\\":\\"Enum1\\",\\"requiredBoolean\\":false,\\"requiredConst\\":\\"some-const-value\\",\\"requiredArray\\":[\\"required\\",\\"array\\"],\\"optionalString\\":\\"I'm optional!\\"}",
  ],
  Array [
    "CustomPlugin: ",
    "validate() event='context' properties={\\"requiredString\\":\\"A required string\\",\\"optionalEnum\\":\\"Value 1\\"}",
  ],
  Array [
    "CustomPlugin: ",
    "track() userId='undefined' event='Event With All Properties' properties={\\"requiredString\\":\\"A required string\\",\\"optionalEnum\\":\\"Value 1\\",\\"requiredNumber\\":2,\\"requiredInteger\\":42,\\"requiredEnum\\":\\"Enum1\\",\\"requiredBoolean\\":false,\\"requiredConst\\":\\"some-const-value\\",\\"requiredArray\\":[\\"required\\",\\"array\\"],\\"optionalString\\":\\"I'm optional!\\"}",
  ],
  Array [
    "CustomPlugin: ",
    "postTrack() userId='undefined' event='Event With All Properties' properties={\\"requiredString\\":\\"A required string\\",\\"optionalEnum\\":\\"Value 1\\",\\"requiredNumber\\":2,\\"requiredInteger\\":42,\\"requiredEnum\\":\\"Enum1\\",\\"requiredBoolean\\":false,\\"requiredConst\\":\\"some-const-value\\",\\"requiredArray\\":[\\"required\\",\\"array\\"],\\"optionalString\\":\\"I'm optional!\\"} validationResponses=[{\\"valid\\":true,\\"pluginId\\":\\"custom\\"},{\\"valid\\":true,\\"pluginId\\":\\"custom\\"}]",
  ],
  Array [
    "CustomPlugin: ",
    "validate() event='EventMaxIntForTest' properties={\\"intMax10\\":20}",
  ],
  Array [
    "CustomPlugin: ",
    "validate() event='context' properties={\\"requiredString\\":\\"A required string\\",\\"optionalEnum\\":\\"Value 1\\"}",
  ],
  Array [
    "CustomPlugin: ",
    "track() userId='undefined' event='EventMaxIntForTest' properties={\\"requiredString\\":\\"A required string\\",\\"optionalEnum\\":\\"Value 1\\",\\"intMax10\\":20}",
  ],
  Array [
    "CustomPlugin: ",
    "postTrack() userId='undefined' event='EventMaxIntForTest' properties={\\"requiredString\\":\\"A required string\\",\\"optionalEnum\\":\\"Value 1\\",\\"intMax10\\":20} validationResponses=[{\\"valid\\":true,\\"pluginId\\":\\"custom\\"},{\\"valid\\":true,\\"pluginId\\":\\"custom\\"}]",
  ],
  Array [
    "CustomPlugin.id()",
    "custom",
  ],
]
`;
